lib/MusicBrainz/Server/Data/Utils.pm

```diff
@@ -254,33 +254,22 @@ sub is_valid_token {
     defined $token && $token =~ /^[A-Za-z0-9_-]+$/;
 }
 
-sub get_area_containment_query {
-    my ($link_type_param, $descendant_param, %args) = @_;
-
-    my $levels_condition = $args{check_all_levels}
-        ? ''
-        : ' JOIN area a ON a.id = ad.parent WHERE a.type IN (1, 2, 3) ';
-
-    return ("
-        WITH RECURSIVE area_descendants AS (
-            SELECT entity0 AS parent, entity1 AS descendant, 1 AS depth
-              FROM l_area_area laa
-              JOIN link ON laa.link = link.id
-             WHERE link.link_type = $link_type_param
-               AND entity1 = $descendant_param
-                UNION
-            SELECT entity0 AS parent, descendant, (depth + 1) AS depth
-              FROM l_area_area laa
-              JOIN link ON laa.link = link.id
-              JOIN area_descendants ON area_descendants.parent = laa.entity1
-             WHERE link.link_type = $link_type_param
-               AND entity0 != descendant
-        )
-        SELECT ad.descendant, ad.parent, ad.depth
-          FROM area_descendants ad
-         $levels_condition
-         ORDER BY ad.descendant, ad.depth ASC",
-        $PART_OF_AREA_LINK_TYPE_ID);
+sub get_area_containment_join {
+    my ($sql) = @_;
+
+    CORE::state $has_materialized_data;
+
+    if (!defined $has_materialized_data) {
+        $has_materialized_data = $sql->select_single_value(
+            'SELECT 1 FROM area_containment LIMIT 1',
+        ) ? 1 : 0;
+    }
+
+    if ($has_materialized_data) {
+        return 'area_containment';
+    }
+
+    return 'get_area_parent_hierarchy_rows(NULL)';
```

Comment:
You've suggested we should be using DISTINCT with `get_area_parent_hierarchy_rows`, but AFAICT you don't when using this as a fallback?

Reply:
Good point!